hypothesis-python/src/hypothesis/internal/conjecture/shrinking/length.py

```diff
@@ -52,50 +53,65 @@ def left_is_better(self, left, right):
         return len(left) < len(right)
 
     def run_step(self):
-        # Try to delete as many elements as possible from the sequence, in
-        # (roughly) one pass, from right to left.
-
-        # Starting from the end of the sequence, we try to delete as many
-        # consecutive elements as possible. When we encounter an element that
-        # can't be deleted this way, we skip over it for the rest of the pass,
-        # and continue to its left. This lets us finish a pass in linear time,
-        # but the drawback is that we'll miss some possible deletions of
-        # already-skipped elements.
-        skipped = 0
-
-        # When every element has been deleted or skipped, the pass is complete.
-        while skipped < len(self.current):
-            # Number of remaining elements to the left of the skipped region.
-            # These are all candidates for attempted deletion.
-            candidates = len(self.current) - skipped
-
-            # Take a stable snapshot of the current sequence, so that deleting
-            # elements doesn't mess with our slice indices.
-            start = self.current
-
-            # Delete as many elements as possible (k) from the candidate
-            # region, from right to left. Always retain the skipped elements
-            # at the end. (See diagram below.)
-            find_integer(
-                lambda k: k <= candidates
-                and self.consider(start[: candidates - k] + start[candidates:])
+        # Try to delete as many elements as possible from the sequence, trying
+        # each element no more than once.
+
+        # We convert the sequence to a set of indices. This allows us to more
+        # easily do book-keeping around which elements we've tried removing.
+        initial = self.current
+
+        indices = list(hrange(len(self.current)))
+
+        # The set of indices that we have so far decided to retain.
```

Comment:
Minor: “Decided to retain” is a bit awkward here, since for many of these indices we haven't made any concrete decision yet.

Suggestion: The set of indices that currently remain, either because we have decided to retain them, or because we haven't tried removing them yet.